### PR TITLE
Move Dygraph Indexing Parsing to Pybind

### PR DESCRIPTION
**Pull request by zoooo0820**
### PR types

Performance optimization
### PR changes

Others
### Description

Pcard-66985 

For now, although there  there is already complete indexing (i.e. `__getitem__` / `__setitem__`) functional support, almost all the indexing parsing parts are running in Python, which brings much performance overhead.

In this PR,  we move indexing parsing of dygraph from Python to Pybind, aiming to optimize the performance (For static mode, still in Python). 

**some other things in this pr:**
- add a fast path for single tensor getting (like `x[tensor[1,2]]` ): by replace the `stack` to `unsqueeze`.
- remove non-necessary `transpose` in advanced getting.
- remove `any` check in bool-mask indexing, since the any kernel and gpu->cpu memcpy is quite slow.
- remove `cond` in bool-mask indexing, since currently this conditional block cannot be parsed in PIR

For convenience of review, the old indexing-parsing code will be deleted later.

### CODE REVIEW

File: paddle/fluid/pybind/slice_utils.h [hunk 1]
```diff
@@ -305,5 +308,364 @@ static void ParseIndexingSlice(phi::DDim shape,
...
+            "The dimension of bool index doesn't match indexed array along "
+            "dimension %d, the target dimension is %d, but received %d",
+            i,
+            tensor_shape[i],
+            bool_index.shape()[i]));
+    i++;
+  }
+
+  auto bool_2_idx = nonzero_ad_func(bool_index);
+  return gather_nd_ad_func(tensor, bool_2_idx);
```
Comment (wanghuancoder) [inline paddle/fluid/pybind/slice_utils.h]: bool_index, tensor这两个都是内部创建的Tensor吗？如果有1个是源自于Python 用户的。就需要做转DistTensor的处理。

File: paddle/fluid/pybind/slice_utils.h [hunk 2]
```diff
@@ -305,5 +308,364 @@ static void ParseIndexingSlice(phi::DDim shape,
...
+}
+
+static void ParseBoolAndBroadcastIndices(
+    std::vector<paddle::Tensor>* advanced_index) {
```
Comment (wanghuancoder) [inline paddle/fluid/pybind/slice_utils.h:661]: advanced_index如果是源自于用户的，Vector里可能有的是Dense有的是Dist需要处理。

File: paddle/fluid/pybind/slice_utils.h [hunk 3]
```diff
@@ -305,5 +308,364 @@ static void ParseIndexingSlice(phi::DDim shape,
...
+  for (size_t i = 0; i < advanced_index->size(); i++) {
+    if ((*advanced_index)[i].dtype() == phi::DataType::BOOL) {
+      paddle::Tensor bool_2_idx = nonzero_ad_func((*advanced_index)[i]);
+      paddle::Tensor bool_2_idx_sliced =
+          slice_ad_func(bool_2_idx, {1}, {0}, {1}, {1}, {1});
+      (*advanced_index)[i] = bool_2_idx_sliced;
```
Comment (wanghuancoder) [inline paddle/fluid/pybind/slice_utils.h:667]: advanced_index， bool_2_idx_sliced如果有1个源自于用户的，则需要做转换。

File: paddle/fluid/pybind/slice_utils.h (hunk 2 above)
Comment (zoooo0820) [reply paddle/fluid/pybind/slice_utils.h:661]: done，已在前一个函数检查和处理

**Review by wanghuancoder — APPROVED** [review]
LGTM

**Review by jeff41404 — APPROVED** [review]
LGTM

**Pull request merged**